gherkin/java/src/main/java/io/cucumber/gherkin/TokenMatcher.java

```diff
@@ -166,29 +173,51 @@ private boolean match_DocStringSeparator(Token token, String separator, boolean
                 indentToRemove = 0;
             }
 
-            setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, null, null);
+            setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, null, null, null);
             return true;
         }
         return false;
     }
 
     @Override
     public boolean match_StepLine(Token token) {
-        List<String> keywords = currentDialect.getStepKeywords();
-        for (String keyword : keywords) {
-            if (token.line.startsWith(keyword)) {
-                String stepText = token.line.getRestTrimmed(keyword.length());
-                setTokenMatched(token, TokenType.StepLine, stepText, keyword, null, null);
-                return true;
+        Map<StepKeywordType, List<String>> keywords = currentDialect.getStepKeywordsByType();
+
+        String keyword = null;
+        StepKeywordType keywordType = StepKeywordType.UNKNOWN;
+        for (Entry<StepKeywordType,List<String>> entry : keywords.entrySet()) {
```

Comment:
The code generates the following mapping:

| step keyword type | prior keyword type | keyword appears in given/when/then translations | output (keywordType)
|--------|-----------|--|--|
| CONJUNCTION | none (initial) | once | UNKNOWN (using `lastKeywordType`)
| CONJUNCTION | `lastKeywordType` | once | `lastKeywordType`
| CONTEXT | (not applicable) | once | CONTEXT
| ACTION  | (not applicable) | once | ACTION
| OUTCOME  | (not applicable) | once | OUTCOME
| CONTEX/ACTION/OUTCOME/CONJUNCTION | (not applicable) | multiple (e.g. given+when) | _**UNKNOWN**_

The last row is the one that relates to your question below. Because there are multiple mappings of a single keyword translation, we cannot determine the exact relation _translated keyword_ -> _Given/When/Then/But/And_; hence the conclusion `unknown`.

Reply:
My last push seriously simplifies the loop. I hope that's an improvement.

Reply:
I feel I have to attach a debugger to understand what is going on. But not having done so yet, my comments:

I'd expect to see to see the resolution of multiple mappings reflected in the code rather then implicitly brushed over. For example:

```
List<Entry<StepKeywordType, String> candidates = keywords.entrySet()
   .filter( ... some condition .... )
   .map( ... some transformation .... )   
   .collect(....);
   
if (candidates.size() != 1)
   return unknown
else
  return candidates.get(0).....
```

Though this also makes me wonder why a `Set<StepKeywordType,List<String>>` is used rather then a `Set<String,List<StepKeywordType>>`. Then the resolution of the keyword type is a lookup rather then an algorithm.

Reply:
Heh. It seems our comments just crossed each other. The new code no longer uses the problematic pattern in this place of the code. I think that makes this bit more readable. A similar pattern is now in `GherkinDialect` though, which precomputes the `Map` that @ciaranmcnulty suggested to use in the `match_StepLine` function. There, I've added a comment a moment ago to indicate why the override on the type is happening.

Reply:
This is now implemented as you suggested.